sklearn/utils/_weight_vector.pyx.tp

```diff
@@ -145,20 +174,20 @@ cdef class WeightVector(object):
 
         Returns
         -------
-        innerprod : double
+        innerprod : {{c_type}}
             The inner product of ``x`` and ``w``.
         """
         cdef int j
         cdef int idx
-        cdef double innerprod = 0.0
-        cdef double* w_data_ptr = self.w_data_ptr
+        cdef {{c_type}} innerprod = 0.0
+        cdef {{c_type}}* w_data_ptr = self.w_data_ptr
         for j in range(xnnz):
             idx = x_ind_ptr[j]
             innerprod += w_data_ptr[idx] * x_data_ptr[j]
         innerprod *= self.wscale
         return innerprod
 
-    cdef void scale(self, double c) nogil:
+    cdef void scale(self, {{c_type}} c) nogil:
```

Comment:
Shall the adapt the epsilon value for the comparision on `self.wscale` (at line 197) based on `c_type`?

Reply:
Indeed, it's probably a good idea to use `1e-7` or even `1e-6` when `self.wscale` has type `float32`. Otherwise I am not sure this condition will even be triggered on float32 data.

Or we could just use `1e-6` for both dtypes. But maybe this will have non-trivial impacts on the performance of `SGDClassifier` and co on `np.float64` which we should try to avoid introducing as part of this PR.

Unfortunately it's not easy to write tests for this because this is a Cython class with `cdef` methods.

Maybe we can set epsilon to `1e-6` for `float32` and keep `1e-9` unchanged for `float64` and just write tests for the float32 case at the public Python API level only in `test_sgd.py` in a subsequent PR where we enable float32 support for linear SGD models.